### PR TITLE
feat: helper contract

### LINKED ISSUE #76: feat: implement helper

### LINKED ISSUE #77: chore: add safe legacy trading params to helper

### PR DESCRIPTION
**Pull request by mfw78**
## Description

This PR provides a helper contract that may be utilised by entities who are indexing `ConstantProduct` CoW AMMs, allowing them to generate a canonical order, corresponding pre-interaction and signature.

In providing the helper contract, a snapshot of legacy (safe-based) CoW AMMs is provided for Ethereum mainnet and Gnosis Chain. Legacy helper logic is separated from the standalone helper contract such to provide as clean as possible a reference implementation for AMMs implementing the `ICOWAMMPoolHelper` interface.

Testing of the helper contract is handled in #83.

## Tasks

- [x] Update factory addresses on deployment to `mainnet` / `gnosis`.

Replaces #72
Closes #76 
Closes #77

### CODE REVIEW

File: src/legacy/Helper.sol [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                token1: _tokens[1],
+                priceNumerator: prices[0],
+                priceDenominator: prices[1],
+                appData: tradingParams.appData
+            })
+        );
+
+        bytes32 domainSeparator = SETTLEMENT.domainSeparator();
+
+        sig = abi.encodeWithSignature(
```
Comment (fleupold) [inline src/legacy/Helper.sol:58]: Do we have an e2e test for this?

File: src/libraries/GetTradeableOrder.sol [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        IERC20 token1;
+        uint256 priceNumerator;
+        uint256 priceDenominator;
+        bytes32 appData;
+    }
+
+    /// @notice The largest possible duration of any AMM order, starting from the current block timestamp.
+    uint32 public constant MAX_ORDER_DURATION = 5 * 60;
+
+    function getTradeableOrder(GetTradeableOrderParams memory params)
```
Comment (fleupold) [inline src/libraries/GetTradeableOrder.sol:21]: Assuming the math in this function was copied from the existing code.

**Review by fleupold — APPROVED** [review]
Looks very good!

File: src/interfaces/ICOWAMMPoolHelper.sol [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+     * @param prices supplied for determining the order, assumed to be in the
+     *        same order as returned from `tokens(pool)`.
+     * @return The GPv2Order.Data struct for the CoW Protocol JIT order, any
+     *          PRE and/or POST interaction that is required by the **SOLVER**, and
+     *          the pool's associated ERC-1271 signature.
+     */
+    function order(address pool, uint256[] calldata prices)
+        external
+        view
+        returns (GPv2Order.Data memory, GPv2Interaction.Data[] memory, bytes memory);
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/interfaces/ICOWAMMPoolHelper.sol:63]: The reason for the array is in the event of needing to pass a pre _and_ post hook.

File: src/libraries/GetTradeableOrder.sol (hunk 2 above)
Comment (mfw78) [reply src/libraries/GetTradeableOrder.sol:21]: Absolutely, with the exclusion of changing out one of the sub commands which was revert guarded to make it play well previously with the watch-tower.

File: src/legacy/Helper.sol (hunk 1 above)
Comment (mfw78) [reply src/legacy/Helper.sol:58]: See #83

File: src/ConstantProductHelper.sol [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import {IERC20, GPv2Order, ISettlement, ConstantProduct} from "./ConstantProduct.sol";
+import {ICOWAMMPoolHelper, GPv2Interaction} from "./interfaces/ICOWAMMPoolHelper.sol";
+import {GetTradeableOrder} from "./libraries/GetTradeableOrder.sol";
+import {ConstantProductFactory} from "./ConstantProductFactory.sol";
+import {Helper as LegacyHelper} from "./legacy/Helper.sol";
+
+contract ConstantProductHelper is ICOWAMMPoolHelper, LegacyHelper {
+    using GPv2Order for GPv2Order.Data;
+
+    address public constant factory = address(0);
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/ConstantProductHelper.sol:13]: Yep, it's in the tasks that it will be hard coded once the factory is deployed on the respective chains.

**Review by fedgiac — COMMENTED** [review]
Looks good for its purpose. The things that need to be addressed in my review before merging are:
- the output type of interactions in `order()`.
- adding post-interaction for legacy safe-based amms.

File: src/interfaces/ICOWAMMPoolHelper.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (fleupold) [reply src/interfaces/ICOWAMMPoolHelper.sol:63]: Agree, without reading the code if I was a developer I'd assume this was all pre interactions.

File: src/legacy/Helper.sol [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            "safeSignature(bytes32,bytes32,bytes,bytes)",
+            domainSeparator,
+            GPv2Order.TYPE_HASH,
+            abi.encode(_order),
+            abi.encode(
+                IExtensibleFallbackHandler.PayloadStruct({params: params, offchainInput: "", proof: new bytes32[](0)})
+            )
+        );
+
+        interactions = new GPv2Interaction.Data[](1);
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/legacy/Helper.sol:68]: > The interface in the specifications requires to use GPv2Interaction.Data[2] instead, where [0] is pre and [1] is post (and didn't specify what to do if no interaction is needed, but we could say that the to address is the zero address).
Using an array creates a lot of implicit assumptions that aren't mentioned in the docs and are very confusing.

The interface was modified for these learnings, and has been applied in commit 2aec80d.

File: src/interfaces/ICOWAMMPoolHelper.sol [hunk 6]
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.24;
+
+import {GPv2Order} from "lib/composable-cow/src/BaseConditionalOrder.sol";
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/interfaces/ICOWAMMPoolHelper.sol:4]: Great point, done.

File: src/interfaces/ICOWAMMPoolHelper.sol [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    function tokens(address pool) external view returns (address[] memory);
+    /**
+     * AMM Pool helpers MUST provide a method for returning the canonical order
+     * required to satisfy the pool's invariants, given a pricing vector.
+     * @dev Reverts with `NoOrder` if the `pool` has no canonical order matching the
+     *      given price vector.
+     * @param pool to calculate the order / signature for
+     * @param prices supplied for determining the order, assumed to be in the
+     *        same order as returned from `tokens(pool)`.
+     * @return The GPv2Order.Data struct for the CoW Protocol JIT order, any
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/interfaces/ICOWAMMPoolHelper.sol]: Adjusted for the revised return types in the interface.

File: src/legacy/Helper.sol [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    function isLegacyEnabled(address amm) public view returns (bool success) {
+        // First check the safe has `ExtensibleFallbackHandler` configured as the fallback handler
+        address fallbackHandler =
+            abi.decode(IStorageViewer(amm).getStorageAt(uint256(FALLBACK_HANDLER_STORAGE_SLOT), 1), (address));
+        success = fallbackHandler == EXTENSIBLE_FALLBACK_HANDLER;
+
+        // Next check that the safe has delegated the domain verifier to ComposableCoW
+        success = success
+            && IExtensibleFallbackHandler(EXTENSIBLE_FALLBACK_HANDLER).domainVerifiers(amm, SETTLEMENT.domainSeparator())
+                == address(COMPOSABLE_COW);
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/legacy/Helper.sol:110]: Done.

File: src/legacy/Helper.sol [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        IConditionalOrder.ConditionalOrderParams params;
+        bytes offchainInput;
+    }
+
+    function domainVerifiers(address safe, bytes32 domainSeparator) external view returns (address);
+}
+
+interface IComposableCOW {
+    function singleOrders(address safe, bytes32 hash) external view returns (bool);
+}
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/legacy/Helper.sol:151]: Was hopeful that we could actually remove composable cow as a dependency altogether, but with the e2e settlement test this doesn't look likely. In which case, sure!

File: src/legacy/Snapshot.sol [hunk 10]
```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: LGPL-3.0-only
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/legacy/Snapshot.sol:1]: Blurb added.

File: src/interfaces/ICOWAMMPoolHelper.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (mfw78) [reply src/interfaces/ICOWAMMPoolHelper.sol]: The interface was modified to break these out into separate pre and post (return tuple of 4 elements)

File: src/ConstantProductHelper.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (mfw78) [reply src/ConstantProductHelper.sol]: Done.

File: src/ConstantProductHelper.sol [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        // Mainnet, gnosis, and sepolia. Addresses taken from networks.json
+        if (chainId == 1) {
+            return 0x8deEd8ED7C5fCB55884f13F121654Bb4bb7c8437;
+        } else if (chainId == 100) {
+            return 0x2AF6C59FC957D4a45ddBBD927fA30f7C5051F583;
+        } else if (chainId == 11155111) {
+            return 0xbD18758055dbe3ed37A2471394559aE97a5Da5c0;
+        }
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/ConstantProductHelper.sol:32]: It could be, but was lazy :sloth:

**Pull request merged**